### ISSUE TITLE
Listening ports used with wrapping factories could log better start messages

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5292]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5292|
|Type|enhancement|
|Created|2011-10-05 16:44:30Z|
|Branch|https://github.com/twisted/twisted/tree/port-log-messages-5292|

When a wrapping factory (`twisted.protocols.policies.WrappingFactory`  - eg `TLSMemoryBIOFactory`, or an endpoints wrapper) is used to listen on a port, the port logs a startup message that refers to the wrapping factory only and not the wrapped (probably equally or more interesting) factory.

Ports are hard-coded to use the factory's `__class__.__name__`.  They could (re-)use `ILoggingContext` to let the factory have some say in this message, and then wrappers could include information from the wrapped factory as well.

(Also, SSL ports sometimes claim they are TCP when they stop.)

<details><summary>Searchable metadata</summary>

```
trac-id__5292 5292
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_port_log_messages_5292 branches/port-log-messages-5292
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1317833070000000 1317833070000000
changetime__1318334494000000 1318334494000000
version__None None
owner__exarkun exarkun
cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [32784]) Branching to 'port-log-messages-5292'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

The port implementations now try to use an `ILoggingContext` implementation if there is one, just as the connection-oriented transports do to get a log prefix.  This lets the wrapping factories include some information about the wrapped factory.

I also fixed the SSL port closed message.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/port-log-messages-5292)

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)|@bitglue set owner to @bitglue|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)<a name="note_4"></a>|@bitglue set owner to @exarkun|
|-|-|

```iocpreactor.tcp.Port._type``` needs minor documentation, like twisted.internet.tcp.Port has. It could just be a comment, but the utility of that variable, and that it's frobbed from listenSSL which is in a different class merits a mention.


I get an error running twisted.internet.test.test_unix.UNIXPortTestsBuilder_SelectReactor.test_startedListeningLogMessage on my OS X box that I don't get on trunk. Not really sure what this is all about:

```
[ERROR]
Traceback (most recent call last):
  File "/Volumes/Macintosh HD/Users/indigo/DevProjects/twisted/port-log-messages-5292/twisted/internet/test/test_tcp.py", line 586, in test_startedListeningLogMessage
    p = self.getListeningPort(reactor, factory)
  File "/Volumes/Macintosh HD/Users/indigo/DevProjects/twisted/port-log-messages-5292/twisted/internet/test/test_unix.py", line 170, in getListeningPort
    return reactor.listenUNIX(self.mktemp()[:-1], factory)
  File "/Volumes/Macintosh HD/Users/indigo/DevProjects/twisted/port-log-messages-5292/twisted/internet/posixbase.py", line 390, in listenUNIX
    p.startListening()
  File "/Volumes/Macintosh HD/Users/indigo/DevProjects/twisted/port-log-messages-5292/twisted/internet/unix.py", line 123, in startListening
    raise CannotListenError, (None, self.port, le)
twisted.internet.error.CannotListenError: Couldn't listen on any:twisted.internet.test.test_unix/UNIXPortTestsBuilder_SelectReact/test_startedListeningLogMessage/9Um4rP/tem: AF_UNIX path too long.
```

How does ```ILoggingContext``` work? I couldn't find any documentation about it. Its
API doc doesn't say anything about how one uses it, it's not mentioned in the logging howto, and it's not referenced from ```twisted.python.log``` anywhere (though it does have ```ILogContext```, which is confusing. Can that documentation be expanded?

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)<a name="note_5"></a>|@bitglue commented|
|-|-|

Oh also, what's with the failures in the build results? I'm not familiar enough with buildbot to really know what's going on there.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun removed owner|
|-|-|

Thanks for the review.  Sorry about the failing tests.  I fixed the `UNIXPortTestsBuilder` failures by switching to `tempfile.mktemp` and asking it to generate names that should be shorter (unix socket names cannot be longer than about 108 bytes - the exact limit is platform dependent).  I fixed the SSL failures by re-arranging the order of inheritance in that test case, so that `TLSMixin` comes first, allowing its `requiredInterfaces` definition to win.

I added a comment for iocpreactor's `_type`, too.

I agree that the documentation for `ILoggingContext` is weak.  I'm a little stumped as to where to put new documentation though.  I don't feel like it's an overall win to add it to the interface API documentation.  I don't think anyone will find it on the implementation API documentation.  It's not about the logging system in general, so it doesn't really belong in the logging howto.  Suggestions welcome.

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)<a name="note_7"></a>|@bitglue commented|
|-|-|

Probably the documentation for the things that use ILoggingContext should be something like this:

```python
def frob(thing):
    """
    Frob a thing.

    C{thing} may provide C{ILoggingContext} to
    customize my logging about frobbing.
    """
```

...and then the ```ILoggingContext``` documentation should give some hints about what sort of things might use providers of ```ILoggingContext``` and why it's useful.

The problem now is that the reader has no way to know why ```ILoggingContext``` is useful or what sorts of things should implement it.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_9"></a>|@jesstess set owner to @exarkun|
|-|-|

Thanks for working on this exarkun, and for the code cleanup along with this commit series. I think this is ready to merge.

As an aside, the UDP logging often doesn't have the transport protocol type as part of the log message where TCP does, e.g. in ```udp.Port.connectionLost``` versus ```tcp.Port._logConnectionLostMsg```.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

Regarding the ILoggingContext documentation, I filed [#5307](https://github.com/twisted/twisted/issues/5307).

Regarding the UDP vs TCP log message inconsistency, I filed [#5308](https://github.com/twisted/twisted/issues/5308).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [32813]) Merge port-log-messages-5292

Author: exarkun
Reviewer: indigo, jesstess
Fixes: [#5292](https://github.com/twisted/twisted/issues/5292)

When logging start and stop messages for listening ports, respect any
`ILoggingContext` implementation provided by the factory and use the
log prefix supplied to identify what is being started or stopped in
the log message.
